File: 10/day10.pl
:- use_module(library(clpfd)).

input(Asteroids) :-
    read_file_to_string("day10.input", String, []),
    split_string(String, "\n", "", SplitRaw),
    select("", SplitRaw, Split),
    grid_to_asteroids(Split, Asteroids).
    
grid_to_asteroids(Grid, Asteroids) :-
    maplist([In, Out]>>(string_chars(In, Chars), enumerate(Chars, Out, 0)), Grid, Xs),
    enumerate(Xs, Ys, 0),
    maplist([Y-List, Out]>>convlist({Y}/[A,B]>>(A=X-'#',B=X-Y), List, Out), Ys, XYs),
    flatten(XYs, Asteroids).

enumerate([], [], _).
enumerate([H|T], [I-H|TT], I) :-
    NI #= I + 1,
    enumerate(T, TT, NI).

gcd(X, Y, GCD) :-
    ( X #= 0
    -> 
        GCD #= abs(Y)
    ;
        Z #= Y mod X,
        gcd(Z, X, GCD)
    ).

collision(Asteroids, FX-FY, TX-TY, DX-DY) :-
    CX #= FX+DX,
    CY #= FY+DY,
    CX-CY \= TX-TY,
    (
        memberchk(CX-CY, Asteroids)
    ;
        collision(Asteroids, CX-CY, TX-TY, DX-DY)
    ).

detect(From, Asteroids, Detected) :-
    detect(From, 0, Asteroids, Detected).

detect(_, Index, Asteroids, 0) :-
    length(Asteroids, Index).

detect(FX-FY, Index, Asteroids, Detected) :-
    length(Asteroids, N),
    Index #< N,
    nth0(Index, Asteroids, TX-TY),
    NI #= Index + 1,
    detect(FX-FY, NI, Asteroids, PrevDetected),
    DiffX #= TX-FX,
    DiffY #= TY-FY,
    gcd(DiffX, DiffY, Step),
    DX #= DiffX // Step,
    DY #= DiffY // Step,
    ( collision(Asteroids, FX-FY, TX-TY, DX-DY)
    ->
        Detected #= PrevDetected
    ;
        Detected #= PrevDetected+1
    ).

part1(Asteroids, Ans, StationX, StationY) :-
    maplist({Asteroids}/[A, D-A]>>(selectchk(A, Asteroids, Rem), detect(A, Rem, D)), Asteroids, List),
    max_member(Ans-(StationX-StationY), List).

part2(Asteroids, SX, SY, Out) :-
    % maplist Asteroids to Order-Asteroid
    % collect Asteroids with same order in one list under key
    % sort the list by Order
    % cycle through the list and consume asteroids
    % return the 200th asteroid consumed
    % TODO:
    Out = 42.

run :-
    input(Asteroids),
    part1(Asteroids, Detected, StationX, StationY),
    format('Part 1: ~w~n', [Detected]),
    part2(Asteroids, StationX, StationY, Ans),
    format('Part 2: ~w~n', [Ans]).

:- begin_tests(part1).

test(simple_grid) :-
    Grid = [".#..#",
            ".....",
            "#####",
            "....#",
            "...##"],
    grid_to_asteroids(Grid, Asteroids),
    part1(Asteroids, D, SX, SY),
    assertion( 8 == D ),
    assertion( SX-SY == 3-4 ).

:- end_tests(part1).

File: 10/day10.rs
use std::fs::File;
use std::io;
use std::io::prelude::*;
use std::collections::HashMap;

fn read_file(filename: &str) -> io::Result<io::BufReader<File>> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    Ok(reader)
}

fn parse(reader: io::BufReader<File>) -> HashMap<(i32,i32), i32> {
    let mut m = HashMap::new();
    for (y, line) in reader.lines().enumerate() {
        let line = line.unwrap();
        for (x, c) in line.chars().enumerate() {
            match c {
                '#' => m.insert((x as i32, y as i32), 0),
                '.' => continue,
                _   => panic!()
            };
        }
    }
    m
}

fn gcd(x: i32, y: i32) -> i32 {
    if x == 0 {
        y.abs()
    } else {
        gcd(y % x, x)
    }
}

fn collision(m: &HashMap<(i32,i32),i32>, from: (i32,i32), to: (i32,i32), steps: (i32,i32)) -> bool {
    let mut coords = from;
    loop {
        let (x, y) = coords;
        let (dx, dy) = steps;
        coords = (x+dx, y+dy);
        if coords == to {
            return false
        }
        let asteroid = m.get(&coords);
        if asteroid != None {
            return true
        }
    }
}

fn main() {
    let reader = read_file("day10.input").unwrap();
    let m = parse(reader);
    let mut max = 0;
    for (from, _) in &m {
        let mut detected = 0;
        for (to, _) in &m {
            if from == to {
                continue
            }
            let (x, y) = from;
            let (x2, y2) = to;
            let (diffx, diffy) = (x2-x, y2-y);
            let steps = gcd(diffx, diffy);
            let (diffx, diffy) = (diffx/steps, diffy/steps);
            if collision(&m, *from, *to, (diffx, diffy)) {
                continue
            }
            detected += 1;
        }
        if max < detected {
            max = detected;
        }
    }
    println!("Part 1: {}", max);
}
